### PR TITLE
NETSCRIPT: Faster API wrapping on script launch.

### PR DESCRIPTION
**Pull request by Snarling**
Credit to @d0sboots , this is basically his work from #223 just refactored a bit.

* ns API is wrapped once
* reference to workerscript is kept in a private class field, and this is all that needs to be updated when running a new script. All layers of ns are stamped with this private class field.
* Significantly increases maximum number of scripts before out-of-memory crash (for performance reasons it is still not recommended to run more than a few thousand scripts at once).
* Significantly speeds up the time needed to generate a new WorkerScript.

**_BREAKING CHANGE:_** if you take an ns function and put it into another variable off of ns, you will need to bind the function to ns (or to its parent layer) for it to continue to work. e.g. 
```js
// No longer works
let myTprint1 = ns.tprint

// Still works
let myTprint2 = ns.tprint.bind(ns);

myTprint2("Hello world"); // Works
myTprint1("Goodbye world"); // Error due to function not having an ns layer as a this value
```

There are still types that should be improved, todos have been added in those locations.

### Benchmark1:
10k script executions, both one-at-a-time and all-at-once. Over 6x faster in one-at-a-time execution and almost 2x as fast for all-at-once.
image

### Benchmark2:
Running as many scripts as possible before browser crashes due to out-of-memory error. Also include how long it took to launch each of the first 60 sets of 1000 scripts.
**Dev**: browser crashed at 63000
image

**This PR**: At least 344k scripts at once. Test had ran for a very long time at this point. so I just closed it.

https://user-images.githubusercontent.com/84951833/204103009-7071bf49-7f6b-4b89-a014-b4e5960291a1.mp4

PR60kScriptTimes

### CODE REVIEW

File: src/NetscriptFunctions.ts [hunk 1]
```diff
@@ -1913,6 +1909,34 @@ export const ns = {
...
+
+export function NetscriptFunctions(ws: WorkerScript): ExternalAPI<NSFull> {
+  const instance = new StampedLayer(ws, wrappedNS) as any;
+  instance.args = ws.args.slice();
+  stampLayers(ws, instance, wrappedNS);
+  return instance;
+}
+
+function stampLayers(ws: WorkerScript, stamped: any, unstamped: any) {
+  for (const [k, v] of Object.entries(unstamped)) {
```
Comment (d0sboots) [inline src/NetscriptFunctions.ts]: Suggestion: We've already figured out the nature of all the keys in wrapAPILayer. If that returned a secondary object that was just a pruned tree of API names, we could traverse just that and avoid checking the whole API tree at script creation time.

File: src/NetscriptFunctions.ts [hunk 2]
```diff
@@ -1913,6 +1909,34 @@ export const ns = {
...
+    if (typeof v === "object" && k !== "enums") {
+      stamped[k] = new StampedLayer(ws, v);
+      stampLayers(ws, stamped[k], unstamped[k]);
```
Comment (d0sboots) [inline src/NetscriptFunctions.ts]: `unstamped[k]` is just `v`.

File: src/NetscriptFunctions.ts [hunk 3]
```diff
@@ -1913,6 +1909,34 @@ export const ns = {
   ...NetscriptExtra(),
 };
 
+const wrappedNS = wrapAPILayer({} as ExternalAPI<NSFull>, ns, []);
+function freezeEnums(obj: object, isEnums = false) {
```
Comment (d0sboots) [inline src/NetscriptFunctions.ts]: Why not do this in wrapAPILayer, which is already doing this traversal *and* checking for enums?

**Review by d0sboots — COMMENTED** [review]
I hope you don't mind me leaving a few comments, since you rewrote my code. :)

File: src/NetscriptFunctions.ts (hunk 3 above)
Comment (Snarling) [reply src/NetscriptFunctions.ts]: Moved enum freezing to wrapAPILayer

File: src/NetscriptFunctions.ts (hunk 1 above)
Comment (Snarling) [reply src/NetscriptFunctions.ts]: Addressed in 1dc7bea

**Pull request merged**